socorro/external/crashstorage_base.py

```diff
@@ -1266,3 +1282,75 @@ def remove(self, crash_id):
             self.tag,
             end_time - start_time
         )
+
+
+class MetricsEnabledBase(RequiredConfig):
+    """Base class for capturing metrics for crashstorage classes"""
+    required_config = Namespace()
+    required_config.add_option(
+        'metrics_prefix',
+        doc='a string to be used as the prefix for metrics keys',
+        default=''
+    )
+    required_config.add_option(
+        'active_list',
+        default='save_raw_and_processed,act',
+        doc='a comma delimeted list of counters that are enabled',
+        from_string_converter=str_to_list
+    )
+
+    def __init__(self, config, namespace='', quit_check_callback=None):
+        self.config = config
+        self.metrics = markus.get_metrics(self.config.metrics_prefix)
+
+    def _make_key(self, *args):
+        return '.'.join(x for x in args if x)
+
+
+class MetricsCounter(MetricsEnabledBase):
+    """Counts the number of times it's called"""
+    required_config = Namespace()
+
+    def __getattr__(self, attr):
+        if attr in self.config.active_list:
+            self.metrics.incr(attr)
+        return self._noop
+
+    def _noop(self, *args, **kwargs):
+        pass
+
+
+class MetricsBenchmarkingWrapper(MetricsEnabledBase):
+    """Sends timings for specified method calls in wrapped crash store"""
+    required_config = Namespace()
+    required_config.add_option(
+        name="wrapped_object_class",
+        doc='fully qualified Python class path for an object to an be benchmarked',
+        default='',
+        from_string_converter=class_converter
+    )
+
+    def __init__(self, config, namespace='', quit_check_callback=None):
+        super(MetricsBenchmarkingWrapper, self).__init__(
+            config,
+            namespace=namespace,
+            quit_check_callback=quit_check_callback
+        )
+        self.wrapped_object = config.wrapped_object_class(
+            config,
+            namespace=namespace,
+            quit_check_callback=quit_check_callback
+        )
+
+    def close(self):
+        self.wrapped_object.close()
+
+    def __getattr__(self, attr):
+        wrapped_attr = getattr(self.wrapped_object, attr)
+        if attr in self.config.active_list:
+            def benchmarker(*args, **kwargs):
+                metrics_key = self._make_key(self.config.wrapped_object_class.__name__, attr)
+                with self.metrics.timer(metrics_key):
+                    return wrapped_attr(*args, **kwargs)
+            return benchmarker
```

Comment:
Yes--you understand correctly. I'm not thrilled about how this was implemented. I think it's an awkward way to do aspect-oriented programming. Part of me thinks we should nix these `MetricsBenchmarkingWrapper` altogether and just do all the work with markus directly rather than have a configurationally-determined wrapper. I decided to push off thinking about that more until later. That's why this class is structurally similar to `StatsdBenchmarkingWrapper` still.

The `active_list` denotes the list of counters and timers that are enabled. In this class, those timers are  triggered by methods of the same name.

I'm not sure I want to rename the configuration option to `benchmarked_method_names`. That doesn't feel better. I think I want to think about this more later and redo it more holistically.